ENGAGEHF/Dashboard/Vitals/VitalsManager.swift

```diff
@@ -0,0 +1,353 @@
+///
+/// Functionality:
+/// - Maintain local, up-to-date arrays of the patients health data
+/// - Convert FHIR quantities to HK Samples
+/// - Manage units of the stored quantities, defaulting to the user's phone settings
+@Observable
+public class VitalsManager: Module, EnvironmentAccessible {
+    enum VitalsError: Error {
+        case invalidConversion
+        case unknownUnit
+        case invalidObservationType
+        case missingField
+    }
+    
+    
+    @ObservationIgnored @Dependency private var configureFirebaseApp: ConfigureFirebaseApp
+    @ObservationIgnored @StandardActor private var standard: ENGAGEHFStandard
+    private let logger = Logger(subsystem: "ENGAGEHF", category: "VitalsManager")
+    
+    private var authStateDidChangeListenerHandle: AuthStateDidChangeListenerHandle?
+    private var snapshotListener: ListenerRegistration?
+    
+    // TODO: Localize the units based on either locale or user preferences in the phone settings
+    public var heartRateHistory: [HKQuantitySample] = []
+    public var bloodPressureHistory: [HKCorrelation] = []
+    public var weightHistory: [HKQuantitySample] = []
+    
+    
+    public var latestHeartRate: HKQuantitySample? {
+        heartRateHistory.max { $0.startDate < $1.startDate }
+    }
+    public var latestBloodPressure: HKCorrelation? {
+        bloodPressureHistory.max { $0.startDate < $1.startDate }
+    }
+    public var latestWeight: HKQuantitySample? {
+        weightHistory.max { $0.startDate < $1.startDate }
+    }
+    
+    
+    /// Call on initial configuration:
+    /// - Add a snapshot listener to the three health data collections
+    /// - Set the default units to the user's localized preferences
+    public func configure() {
+        if ProcessInfo.processInfo.isPreviewSimulator {
+            self.setupPreview()
+            return
+        }
+        
+        authStateDidChangeListenerHandle = Auth.auth().addStateDidChangeListener { [weak self] _, user in
+            self?.registerSnapshotListeners(user: user)
+        }
+        
+        self.registerSnapshotListeners(user: Auth.auth().currentUser)
+    }
+    
+    private func registerSnapshotListeners(user: User?) {
+        self.logger.debug("Initializing vitals snapshot listener...")
+        
+        // Remove previous snapshot listener for the user before creating new one
+        snapshotListener?.remove()
+        guard let uid = user?.uid else {
+            self.logger.debug("No user signed in, skipping snapshot listener.")
+            return
+        }
+        
+        let firestore = Firestore.firestore()
+        let userDocRef = firestore.collection("users").document(uid)
+        
+        self.registerWeightSnapshot(userDocRef: userDocRef)
+        self.registerHeartRateSnapshot(userDocRef: userDocRef)
+        self.registerBloodPressureSnapshot(userDocRef: userDocRef)
+    }
+    
+    private func registerWeightSnapshot(userDocRef: DocumentReference) {
```

Comment:
The functionality here is very much duplicated across the two implementations. It might make sense to have a generic function that gets the collection string as an argument and then call this two times with different arguments?